pgjdbc/src/main/java/org/postgresql/core/v3/ConnectionFactoryImpl.java

```diff
@@ -907,36 +914,47 @@ private static void doAuthentication(PGStream pgStream, String host, String user
 
   private static void runInitialQueries(QueryExecutor queryExecutor, Properties info)
       throws SQLException {
-    String assumeMinServerVersion = PGProperty.ASSUME_MIN_SERVER_VERSION.getOrDefault(info);
-    if (Utils.parseServerVersionStr(assumeMinServerVersion) >= ServerVersion.v9_0.getVersionNum()) {
-      // We already sent the parameter values in the StartupMessage so skip this
-      return;
-    }
 
+    // The version we assumed the server would be prior to connecting, to determine what we have already sent
+    Version assumeVersion = ServerVersion.from(PGProperty.ASSUME_MIN_SERVER_VERSION.getOrDefault(info));
+    // The actual version we connected to
     final int dbVersion = queryExecutor.getServerVersionNum();
 
-    if (PGProperty.GROUP_STARTUP_PARAMETERS.getBoolean(info) && dbVersion >= ServerVersion.v9_0.getVersionNum()) {
-      SetupQueryRunner.run(queryExecutor, "BEGIN", false);
-    }
+    // Only need to send the application name if it's defined and wasn't already sent as a startup parameter
```

Comment:
I wouldn't mind seeing some more explanation as to how we used the following tests to come to that conclusion.